sc-memory/sc-memory/sc_agent_context.hpp

```diff
@@ -0,0 +1,338 @@
+      std::function<void(void)> const & initiateCallback,
+      std::function<bool(TScEvent const &)> const & checkCallback) noexcept(false);
+
+  /*!
+   * @brief Creates sc-event wait with condition for the specified sc-event class and subscription sc-element.
+   *
+   * This function allows user to create a wait mechanism that will be triggered
+   * when sc-event of the specified class occurs, provided that the specified condition
+   * is met. This version does not include an initiate callback.
+   *
+   * @param eventClassAddr An address of sc-event class to wait for.
+   *                       This must be a valid sc-element of type `sc_event`.
+   * @param subscriptionElementAddr An address of subscription sc-element, which must be a valid sc-element.
+   * @param checkCallback A callback function that will be called to check if condition is met.
+   *                      It takes a const reference to `ScElementaryEvent` object as a parameter and
+   *                      returns an `bool`.
+   * @return A shared pointer to created `ScWaiter`.
+   * @throws utils::ExceptionInvalidParams If sc-event class or subscription sc-element is not valid.
+   */
+  std::shared_ptr<ScWaiter> CreateConditionWaiter(
+      ScAddr const & eventClassAddr,
+      ScAddr const & subscriptionElementAddr,
+      std::function<bool(ScElementaryEvent const &)> const & checkCallback) noexcept(false);
+
+  /*!
+   * @brief Creates sc-event wait with condition for the specified subscription sc-element and check callback.
+   * @tparam TScEvent A type of sc-event. It must be derived from ScElementaryEvent.
+   * @param subscriptionElementAddr A address of subscription sc-element.
+   * @param checkCallback A callback function that will be called to check if the condition is met. It takes a const
+   * reference to TScEvent as a parameter and returns an bool.
+   * @return A shared pointer to created `ScWaiter`.
+   * @throws utils::ExceptionInvalidParams If subscription sc-element is not valid or if TScEvent is not derived from
+   * ScElementaryEvent.
+   */
+  template <class TScEvent>
+  std::shared_ptr<ScWaiter> CreateConditionWaiter(
+      ScAddr const & subscriptionElementAddr,
+      std::function<bool(TScEvent const &)> const & checkCallback) noexcept(false);
+
+  /*!
+   * @brief Subscribes agent class to specified sc-events.
+   * @tparam TScAgent An agent class to be subscribed from the event.
+   * @param subscriptionAddrs A list of sc-addresses of sc-elements to subscribe to.
+   * @warning Specified agent class must be derived from class `ScAgent`.
+   * @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is already subscribed to the event.
+   */
+  template <class TScAgent, class... TScAddr>
+  _SC_EXTERN typename std::enable_if<!std::is_base_of<ScActionInitiatedAgent, TScAgent>::value>::type SubscribeAgent(
```

Comment:
```suggestion
  _SC_EXTERN typename std::enable_if<!std::is_base_of<ScActionInitiatedAgent, TScAgent>::value>::type SubscribeAgent(
```
I don't like that user can see mention of ScActionInitiatedAgent in context interface. Every time the user wants to look at the method description through the code, he will see this. And it looks like a crutch

Reply:
Let's discuss

Reply:
Fixed